src/devices/sound/discrete.cpp

```diff
@@ -91,61 +91,70 @@ DEFINE_DEVICE_TYPE(DISCRETE, discrete_sound_device, "discrete", "Discrete Sound"
 
 struct output_buffer
 {
+	output_buffer() : source(nullptr), ptr(nullptr)
+	{
+	}
+
+	// required for use in vector (std::atomic<T> has deleted copy constructor)
+	output_buffer(output_buffer &&that) : node_buf(std::move(that.node_buf)), source(that.source), ptr(that.ptr.load(std::memory_order_relaxed)), node_num(that.node_num)
+	{
+	}
+
 	std::unique_ptr<double []>  node_buf;
-	const double                *source;
-	volatile double             *ptr;
+	const double *              source;
+	std::atomic<double *>       ptr;
 	int                         node_num;
 };
 
 struct input_buffer
 {
-	volatile const double       *ptr;               /* pointer into linked_outbuf.nodebuf */
-	output_buffer *             linked_outbuf;      /* what output are we connected to ? */
-	double                      buffer;             /* input[] will point here */
+	const double *              ptr;                // pointer into linked_outbuf.nodebuf
+	output_buffer *             linked_outbuf;      // what output are we connected to?
+	double                      buffer;             // input[] will point here
 };
 
 class discrete_task
 {
-	friend class discrete_device;
 public:
-	virtual ~discrete_task() { }
```

Comment:
Using `public`/`protected`/`private` randomly and relying on friends is a bad design.  (There isn’t much point using `protected` at all here when there are no derived classes.)

There’s no point giving this a vtable as it has no other virtual member functions, and isn’t used as a base class for anything.